### PR TITLE
git xet migrate bugfixes

### PR DESCRIPTION
**Pull request by hoytak**
Fixes many previously unknown issues with git xet repo migrate.
-> Notes were not migrated.
-> Tags were not correctly migrated.
-> Merge and cherry-pick commits were not properly tracked as such. 

The logic to properly do all of this required rethinking the flow.  The new design is MUCH better and more laid out, but is essentially a rewrite of all of it.

### CODE REVIEW

File: rust/gitxetcore/src/git_integration/migration/data_import.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    // Convert &[u8] to a &str to parse as text.
+    let Ok(text) = std::str::from_utf8(data) else {
+        return false;
+    };
+
+    let text = text.trim_start();
+
+    text.starts_with("version https://git-lfs.github.com/spec/v1") && text.contains("oid sha256:")
```
Comment (assafvayner) [inline rust/gitxetcore/src/git_integration/migration/data_import.rs:131]: Are there other headers for lfs pointer files? e.g. for an older spec, or without the github.com title?

Comment (hoytak) [reply rust/gitxetcore/src/git_integration/migration/data_import.rs]: Nope, this is it.

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+
+            // Skip importing of xet notes; we assume that we're rebuilding things.
+            if reference.is_note() && reference_name.starts_with("refs/notes/xet/") {
+                mg_trace!("Xet Note, rejecting.");
+                continue;
+            }
+
+            // Only convert local references
+            if reference.is_remote() {
+                mg_trace!("Xet Note, rejecting.");
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:635]: wrong message.

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+                    if converting_notes {
+                        get_note_tree_dependents(&src, obj, &full_tr_map)
+                    } else {
+                        get_nonnote_tree_dependents(obj)
+                    }
+                }
+                ObjectType::Commit => {
+                    let (dependents, tree_oid) = get_commit_dependents(&src, obj);
+                    if !converting_notes {
+                        // Commits reference the root trees.
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:699]: Add note -- these need to be handled separately so we know to replace or add the .gitattributes file.

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+
+    let Some(tree) = obj.as_tree() else {
+        mg_warn!("Tree Oid {oid} not actually accessible as tree, ignoring.");
+        return vec![];
+    };
+    mg_trace!("Dependences of Note Tree {oid}:");
+
+    for entry in tree.iter() {
+        mg_trace!(" -> Dep: {}", entry.id());
+        dependents.push(entry.id());
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:190]: Add comment -- is note blob containing note text

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+    while let Some(res) = blob_processing_pool.join_next().await {
+        let (b_oid, new_data) = res??;
+        let new_id = dest.blob(&new_data[..])?;
+
+        mg_trace!("BlobTR: {b_oid} -> {new_id}");
+        tr_table.push((b_oid, new_id));
+        progress_reporting.register_progress(Some(1), None);
+    }
+
+    pft.finalize().await?;
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:551]: Remove; happens later.

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 6]
```diff
@@ -0,0 +1,996 @@
...
+    }
+
+    // Now, clear out the rest.
+    while let Some(res) = blob_processing_pool.join_next().await {
+        let (b_oid, new_data) = res??;
+        let new_id = dest.blob(&new_data[..])?;
+
+        mg_trace!("BlobTR: {b_oid} -> {new_id}");
+        tr_table.push((b_oid, new_id));
+        progress_reporting.register_progress(Some(1), None);
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:548]: add data size here.

File: rust/gitxetcore/src/git_integration/repo_migration/migration.rs [hunk 7]
```diff
@@ -0,0 +1,996 @@
...
+    }
+
+    pft.finalize().await?;
+    progress_reporting.finalize();
```
Comment (hoytak) [inline rust/gitxetcore/src/git_integration/repo_migration/migration.rs:552]: Remove, happens later.

File: rust/gitxetcore/src/command/repo.rs [hunk 8]
```diff
@@ -246,6 +246,34 @@ async fn migrate_command(config: XetConfig, args: &MigrateArgs) -> Result<()> {
...
+    let _ = run_git_passthrough(
+        Some(&dest_dir),
+        "push",
+        &["origin", "--force", "--tags", "--follow-tags"],
+        true,
+        false,
+        Some(&[("XET_DISABLE_HOOKS", "0")]),
+    ).map_err(|e| {
+        eprintln!("Error pushing to remote.");
+        eprintln!("Please go to directory {dest_dir:?} and run `git push origin --force --tags --follow-tags` to push manually.");
```
Comment (seanses) [inline rust/gitxetcore/src/command/repo.rs:259]: Re-examine the above pushes, if any error there users wouldn't know to do this push and the one below. Perhaps not to re-throw the errors and continue anyway, if there's no dependency.

**Review by seanses — APPROVED** [review]

**Pull request merged**